charmcraft/format.py

```diff
@@ -0,0 +1,105 @@
+# Copyright 2023 Canonical Ltd.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+# For further info, check https://github.com/canonical/charmcraft
+
+"""Reformat pydantic errors."""
+
+from typing import Tuple
+
+
+def format_pydantic_error_location(loc):
+    """Format location."""
+    loc_parts = []
+    for loc_part in loc:
+        if isinstance(loc_part, str):
+            loc_parts.append(loc_part)
+        elif isinstance(loc_part, int):
+            # Integer indicates an index. Go
+            # back and fix up previous part.
+            previous_part = loc_parts.pop()
+            previous_part += f"[{loc_part}]"
+            loc_parts.append(previous_part)
+        else:
+            raise RuntimeError(f"unhandled loc: {loc_part}")
+
+    loc = ".".join(loc_parts)
+
+    # Filter out internal __root__ detail.
+    loc = loc.replace(".__root__", "")
+    return loc
+
+
+def format_pydantic_error_message(msg):
```

Comment:
```suggestion
def format_pydantic_error_message(msg: str) -> str:
```

Since this is moving anyway :smiley: